### ISSUE TITLE
mint-wallet: Database replaced element! This should not happen!

### ISSUE THREAD

**Issue by Maan2003** [fedimint/minimint#112]
When running the `mint_client::wallet::tests::create_output` unit test with `--nocapture`, this is one of the tracing errors.

**Comment by Maan2003** [comment]
on some debugging

<details>
<summary>the problematic DbBatch</summary>

```rust
{
  BatchItem::DeleteElement(UnsignedTransactionKey (
        Txid(f686284cff0d418fbe3e2bd42d0403a7783a24df399c7bbdf501d36be5da22)
      )),
  BatchItem::InsertNewElement(Element {
      key: PendingTransactionKey (Txid(f686284cff0d418fbe3e2bd42d0403a7783a24df399c7bbdf501d36be5da22)),
      value: PendingTransaction {
          tx: bitcoin::blockdata::transaction::Transaction {
            version: 2,
            lock_time: 0,
            input: Vec(size=1) {
              bitcoin::blockdata::transaction::TxIn {
                previous_output: bitcoin::blockdata::transaction::OutPoint {
                  txid: Txid(00000000000000000000000000000000000000000000000000000000000000),
                  vout: 4294967295
                },
                script_sig: bitcoin::blockdata::script::Script (
                  box
                ),
                sequence: 4294967295,
                witness: Vec(size=5) {
                  Vec(size=0),
                  Vec(size=72) { 48, 69, 2, 33, 0, 155, 165, 66, 65, 111, 82, 178, 234, 225, 150, 149, 75, 238, 89, 18, 20, 228, 88, 71, 202, 16, 38, 155, 76, 0, 163, 2, 219, 154, 168, 239, 127, 2, 32, 102, 95, 235, 155, 239, 39, 197, 210, 129, 226, 182, 126, 233, 178, 92, 4, 0, 63, 231, 14, 224, 1, 123, 38, 239, 197, 142, 17, 0, 129, 200, 22, 1 },
                  Vec(size=72) { 48, 69, 2, 33, 0, 254, 232, 221, 49, 160, 187, 55, 92, 76, 222, 216, 136, 162, 182, 22, 135, 31, 175, 20, 242, 70, 255, 85, 30, 95, 54, 153, 3, 43, 49, 100, 27, 2, 32, 54, 191, 205, 24, 183, 33, 159, 234, 122, 221, 197, 155, 216, 121, 31, 37, 133, 166, 139, 220, 52, 96, 21, 153, 66, 69, 167, 97, 61, 56, 172, 223, 1 },
                  Vec(size=72) { 48, 69, 2, 33, 0, 184, 68, 227, 241, 97, 239, 113, 86, 11, 224, 210, 233, 64, 134, 227, 211, 183, 71, 103, 124, 44, 105, 82, 246, 136, 84, 248, 218, 24, 176, 159, 78, 2, 32, 100, 217, 230, 136, 136, 78, 81, 26, 178, 33, 57, 27, 208, 97, 32, 105, 47, 166, 116, 172, 136, 179, 171, 20, 234, 183, 221, 253, 49, 153, 63, 254, 1 },
                  Vec(size=139) { 83, 33, 2, 41, 85, 97, 120, 164, 122, 74, 25, 182, 201, 19, 84, 254, 251, 111, 197, 58, 251, 210, 123, 48, 93, 185, 211, 39, 145, 90, 234, 49, 221, 9, 254, 33, 2, 66, 143, 236, 32, 91, 140, 244, 9, 15, 85, 42, 246, 125, 70, 208, 95, 80, 42, 56, 99, 130, 33, 246, 154, 207, 107, 71, 103, 169, 145, 24, 113, 33, 3, 113, 209, 74, 172, 166, 119, 156, 177, 127, 179, 15, 53, 62, 62, 249, 73, 14, 57, 84, 172, 107, 61, 218, 141, 197, 212, 131, 208, 48, 110, 60, 202, 33, 3, 214, 6, 53, 115, 53, 239, 88, 94, 59, 206, 91, 237, 91, 22, 83, 120, 231, 39, 50, 108, 228, 181, 70, 172, 134, 165, 102, 166, 154, 174, 21, 91, 84, 174 }
                }
              }
            },
            output: Vec(size=2) {
              bitcoin::blockdata::transaction::TxOut {
                value: 42000,
                script_pubkey: bitcoin::blockdata::script::Script ( { 118, 169, 20, 128, 169, 246, 124, 135, 190, 155, 22, 101, 130, 215, 181, 245, 202, 6, 144, 15, 58, 219, 95, 136, 172 } )
              },
              bitcoin::blockdata::transaction::TxOut {
                value: 4242,
                script_pubkey: bitcoin::blockdata::script::Script ( { 0, 32, 108, 33, 172, 23, 16, 10, 242, 174, 14, 22, 92, 143, 10, 248, 56, 239, 210, 213, 13, 37, 178, 148, 229, 97, 43, 232, 224, 9, 53, 64, 201, 240 } )
              }
            }
          },
          tweak: [ 108, 249, 100, 65, 186, 234, 121, 111, 99, 63, 156, 40, 238, 136, 109, 86, 173, 238, 239, 36, 152, 12, 166, 63, 97, 206, 172, 169, 37, 143, 183, 5 ]
        }
    }),
  BatchItem::DeleteElement(UnsignedTransactionKey (
        Txid(f686284cff0d418fbe3e2bd42d0403a7783a24df399c7bbdf501d36be5da22)
      )),
  BatchItem::InsertNewElement(Element {
      key: PendingTransactionKey (
          Txid(f686284cff0d418fbe3e2bd42d0403a7783a24df399c7bbdf501d36be5da22)
        ),
      value: PendingTransaction {
          tx: bitcoin::blockdata::transaction::Transaction {
            version: 2,
            lock_time: 0,
            input: Vec(size=1) {
              bitcoin::blockdata::transaction::TxIn {
                previous_output: bitcoin::blockdata::transaction::OutPoint {
                  txid: Txid(00000000000000000000000000000000000000000000000000000000000000),
                  vout: 4294967295
                },
                script_sig: bitcoin::blockdata::script::Script (
                  box
                ),
                sequence: 4294967295,
                witness: Vec(size=5) {
                  Vec(size=0),
                  Vec(size=71) { 48, 68, 2, 32, 117, 115, 93, 228, 113, 127, 61, 150, 224, 255, 165, 153, 127, 89, 82, 141, 196, 99, 115, 255, 23, 4, 233, 1, 26, 145, 152, 115, 98, 136, 30, 176, 2, 32, 52, 17, 6, 212, 197, 223, 226, 247, 153, 35, 12, 102, 117, 112, 246, 97, 153, 238, 128, 163, 49, 50, 69, 250, 134, 38, 221, 184, 127, 253, 225, 44, 1 },
                  Vec(size=72) { 48, 69, 2, 33, 0, 155, 165, 66, 65, 111, 82, 178, 234, 225, 150, 149, 75, 238, 89, 18, 20, 228, 88, 71, 202, 16, 38, 155, 76, 0, 163, 2, 219, 154, 168, 239, 127, 2, 32, 102, 95, 235, 155, 239, 39, 197, 210, 129, 226, 182, 126, 233, 178, 92, 4, 0, 63, 231, 14, 224, 1, 123, 38, 239, 197, 142, 17, 0, 129, 200, 22, 1 },
                  Vec(size=72) { 48, 69, 2, 33, 0, 254, 232, 221, 49, 160, 187, 55, 92, 76, 222, 216, 136, 162, 182, 22, 135, 31, 175, 20, 242, 70, 255, 85, 30, 95, 54, 153, 3, 43, 49, 100, 27, 2, 32, 54, 191, 205, 24, 183, 33, 159, 234, 122, 221, 197, 155, 216, 121, 31, 37, 133, 166, 139, 220, 52, 96, 21, 153, 66, 69, 167, 97, 61, 56, 172, 223, 1 },
                  Vec(size=139) { 83, 33, 2, 41, 85, 97, 120, 164, 122, 74, 25, 182, 201, 19, 84, 254, 251, 111, 197, 58, 251, 210, 123, 48, 93, 185, 211, 39, 145, 90, 234, 49, 221, 9, 254, 33, 2, 66, 143, 236, 32, 91, 140, 244, 9, 15, 85, 42, 246, 125, 70, 208, 95, 80, 42, 56, 99, 130, 33, 246, 154, 207, 107, 71, 103, 169, 145, 24, 113, 33, 3, 113, 209, 74, 172, 166, 119, 156, 177, 127, 179, 15, 53, 62, 62, 249, 73, 14, 57, 84, 172, 107, 61, 218, 141, 197, 212, 131, 208, 48, 110, 60, 202, 33, 3, 214, 6, 53, 115, 53, 239, 88, 94, 59, 206, 91, 237, 91, 22, 83, 120, 231, 39, 50, 108, 228, 181, 70, 172, 134, 165, 102, 166, 154, 174, 21, 91, 84, 174 }
                }
              }
            },
            output: Vec(size=2) {
              bitcoin::blockdata::transaction::TxOut {
                value: 42000,
                script_pubkey: bitcoin::blockdata::script::Script ( { 118, 169, 20, 128, 169, 246, 124, 135, 190, 155, 22, 101, 130, 215, 181, 245, 202, 6, 144, 15, 58, 219, 95, 136, 172 })
              },
              bitcoin::blockdata::transaction::TxOut {
                value: 4242,
                script_pubkey: bitcoin::blockdata::script::Script ( { 0, 32, 108, 33, 172, 23, 16, 10, 242, 174, 14, 22, 92, 143, 10, 248, 56, 239, 210, 213, 13, 37, 178, 148, 229, 97, 43, 232, 224, 9, 53, 64, 201, 240 })
              }
            }
          },
          tweak: [ 108, 249, 100, 65, 186, 234, 121, 111, 99, 63, 156, 40, 238, 136, 109, 86, 173, 238, 239, 36, 152, 12, 166, 63, 97, 206, 172, 169, 37, 143, 183, 5 ]
        }
    }),
  BatchItem::InsertElement(Element {
      key: RoundConsensusKey,
      value: minimint_wallet::RoundConsensus {
          block_height: 191,
          fee_rate: minimint_wallet::Feerate {
            sats_per_kvb: 2000
          },
          randomness_beacon: [ 71, 124, 138, 248, 130, 11, 173, 163, 133, 234, 116, 210, 116, 191, 1, 67, 107, 158, 146, 207, 30, 141, 42, 227, 15, 73, 226, 177, 203, 5, 36, 207 ]
        }
    })
}
```
</details>

-  **The first and third items are same.** which leads to
`ERROR Database deleted absent element! This should not happen!`


- **The second and fourth elements are have same key.** which leads to 
`ERROR Database replaced element! This should not happen!`

<br>

The DbBatch is being populated in `mint_wallet::Wallet::begin_consensus_epoch` 
<details>
<summary>consensus_items: </summary>

```rust
Vec(size=8) {
  (
    PeerId(0),
    minimint_wallet::WalletConsensusItem::PegOutSignature(minimint_wallet::PegOutSignatureItem {
        txid: Txid(f686284cff0d418fbe3e2bd42d0403a7783a24df399c7bbdf501d36be5da22),
        signature: Vec(size=1) {
          secp256k1::Signature ( secp256k1_sys::Signature ( [ 78, 159, 176, 24, 218, 248, 84, 136, 246, 82, 105, 44, 124, 103, 71, 183, 211, 227, 134, 64, 233, 210, 224, 11, 86, 113, 239, 97, 241, 227, 68, 184, 254, 63, 153, 49, 253, 221, 183, 234, 20, 171, 179, 136, 172, 116, 166, 47, 105, 32, 97, 208, 27, 57, 33, 178, 26, 81, 78, 136, 136, 230, 217, 100 ]))
        }
      })
  ),
  (
    PeerId(0),
    minimint_wallet::WalletConsensusItem::RoundConsensus(minimint_wallet::RoundConsensusItem {
        block_height: 191,
        fee_rate: minimint_wallet::Feerate {
          sats_per_kvb: 2000
        },
        randomness: [ 24, 194, 16, 237, 214, 45, 145, 240, 49, 202, 20, 192, 125, 254, 3, 241, 87, 106, 153, 61, 59, 14, 221, 144, 214, 24, 245, 55, 87, 33, 164, 38 ]
      })
  ),
  (
    PeerId(1),
    minimint_wallet::WalletConsensusItem::PegOutSignature(minimint_wallet::PegOutSignatureItem {
        txid: Txid(f686284cff0d418fbe3e2bd42d0403a7783a24df399c7bbdf501d36be5da22),
        signature: Vec(size=1) {
          secp256k1::Signature ( secp256k1_sys::Signature ( [ 176, 30, 136, 98, 115, 152, 145, 26, 1, 233, 4, 23, 255, 115, 99, 196, 141, 82, 89, 127, 153, 165, 255, 224, 150, 61, 127, 113, 228, 93, 115, 117, 44, 225, 253, 127, 184, 221, 38, 134, 250, 69, 50, 49, 163, 128, 238, 153, 97, 246, 112, 117, 102, 12, 35, 153, 247, 226, 223, 197, 212, 6, 17, 52 ]))
        }
      })
  ),
  (
    PeerId(1),
    minimint_wallet::WalletConsensusItem::RoundConsensus(minimint_wallet::RoundConsensusItem {
        block_height: 191,
        fee_rate: minimint_wallet::Feerate {
          sats_per_kvb: 2000
        },
        randomness: [ 33, 78, 85, 16, 68, 161, 40, 204, 86, 188, 23, 31, 93, 24, 73, 105, 94, 205, 128, 63, 156, 66, 30, 202, 110, 26, 236, 50, 214, 153, 63, 206 ]
      })
  ),
  (
    PeerId(2),
    minimint_wallet::WalletConsensusItem::PegOutSignature(minimint_wallet::PegOutSignatureItem {
        txid: Txid(f686284cff0d418fbe3e2bd42d0403a7783a24df399c7bbdf501d36be5da22),
        signature: Vec(size=1) {
          secp256k1::Signature ( secp256k1_sys::Signature ( [ 27, 100, 49, 43, 3, 153, 54, 95, 30, 85, 255, 70, 242, 20, 175, 31, 135, 22, 182, 162, 136, 216, 222, 76, 92, 55, 187, 160, 49, 221, 232, 254, 223, 172, 56, 61, 97, 167, 69, 66, 153, 21, 96, 52, 220, 139, 166, 133, 37, 31, 121, 216, 155, 197, 221, 122, 234, 159, 33, 183, 24, 205, 191, 54 ]))
        }
      })
  ),
  (
    PeerId(2),
    minimint_wallet::WalletConsensusItem::RoundConsensus(minimint_wallet::RoundConsensusItem {
        block_height: 191,
        fee_rate: minimint_wallet::Feerate {
          sats_per_kvb: 2000
        },
        randomness: [ 57, 137, 8, 126, 51, 102, 33, 14, 227, 19, 114, 239, 226, 23, 6, 231, 205, 1, 154, 151, 104, 122, 220, 207, 189, 231, 125, 144, 31, 253, 75, 108 ]
      })
  ),
  (
    PeerId(3),
    minimint_wallet::WalletConsensusItem::PegOutSignature(minimint_wallet::PegOutSignatureItem {
        txid: Txid(f686284cff0d418fbe3e2bd42d0403a7783a24df399c7bbdf501d36be5da22),
        signature: Vec(size=1) {
          secp256k1::Signature ( secp256k1_sys::Signature ( [ 127, 239, 168, 154, 219, 2, 163, 0, 76, 155, 38, 16, 202, 71, 88, 228, 20, 18, 89, 238, 75, 149, 150, 225, 234, 178, 82, 111, 65, 66, 165, 155, 22, 200, 129, 0, 17, 142, 197, 239, 38, 123, 1, 224, 14, 231, 63, 0, 4, 92, 178, 233, 126, 182, 226, 129, 210, 197, 39, 239, 155, 235, 95, 102 ]))
        }
      })
  ),
  (
    PeerId(3),
    minimint_wallet::WalletConsensusItem::RoundConsensus(minimint_wallet::RoundConsensusItem {
        block_height: 191,
        fee_rate: minimint_wallet::Feerate {
          sats_per_kvb: 2000
        },
        randomness: [ 71, 121, 199, 123, 35, 225, 53, 145, 1, 143, 5, 226, 182, 78, 77, 60, 175, 56, 17, 90, 209, 187, 53, 118, 10, 172, 134, 36, 85, 64, 244, 75 ]
      })
  )
}
```

</details>

The [faulty Insertions and Deletions](https://github.com/fedimint/minimint/blob/0bbd497a9dc3a5fe164062358b098e9868d111c4/modules/minimint-wallet/src/lib.rs#L630-L631) are done in `Wallet::process_peg_out_signature` 

<details>

<summary>Backtrace</summary>

```
#0  minimint_api::db::batch::AccumulatorTx<minimint_api::db::batch::BatchItem>::append_insert_new<minimint_wallet::db::PendingTransactionKey, minimint_wallet::db::PendingTransaction> (self=*mut ..., key=..., value=...) at /home/manmeet/sr
c/minimint/minimint-api/src/db/batch.rs:197
#1  0x000055f27c81c776 in minimint_wallet::Wallet::process_peg_out_signature (self=*mut ..., batch=..., peer=PeerId(0), signature=*mut ...) at /home/manmeet/src/minimint/modules/minimint-wallet/src/lib.rs:631
#2  0x000055f27c5a17cc in minimint_wallet::{impl#65}::begin_consensus_epoch::{async_block#0}<&mut rand_os::OsRng> () at /home/manmeet/src/minimint/modules/minimint-wallet/src/lib.rs:215
#3  0x000055f27c4bb9ab in core::future::from_generator::{impl#1}::poll<minimint_wallet::{impl#65}::begin_consensus_epoch::{async_block_env#0}<&mut rand_os::OsRng>> (self=..., cx=*mut ...) at /rustc/f1f721e64014863f41c1a386b04af04c2de25321
/library/core/src/future/mod.rs:91
#4  0x000055f27c4a3775 in core::future::future::{impl#1}::poll<alloc::boxed::Box<dyn core::future::future::Future<Output=()>, alloc::alloc::Global>> (self=..., cx=*mut ...) at /rustc/f1f721e64014863f41c1a386b04af04c2de25321/library/core/s
rc/future/future.rs:124
#5  0x000055f27c56e211 in minimint_api::module::testing::{impl#0}::consensus_round::{async_fn#0}<minimint_wallet::Wallet, minimint_wallet::config::WalletClientConfig> () at /home/manmeet/src/minimint/minimint-api/src/module/testing.rs:116
#6  0x000055f27c4bbd8b in core::future::from_generator::{impl#1}::poll<minimint_api::module::testing::{impl#0}::consensus_round::{async_fn_env#0}<minimint_wallet::Wallet, minimint_wallet::config::WalletClientConfig>> (self=..., cx=*mut ..
.) at /rustc/f1f721e64014863f41c1a386b04af04c2de25321/library/core/src/future/mod.rs:91
#7  0x000055f27c574e22 in mint_client::wallet::tests::create_output::{async_block#0}::test_impl::{async_fn#0} () at /home/manmeet/src/minimint/mint-client/src/wallet/mod.rs:283
#8  0x000055f27c4b969b in core::future::from_generator::{impl#1}::poll<mint_client::wallet::tests::create_output::{async_block#0}::test_impl::{async_fn_env#0}> (self=..., cx=*mut ...) at /rustc/f1f721e64014863f41c1a386b04af04c2de25321/lib
rary/core/src/future/mod.rs:91
#9  0x000055f27c572e9e in mint_client::wallet::tests::create_output::{async_block#0} () at /home/manmeet/src/minimint/mint-client/src/wallet/mod.rs:231
#10 0x000055f27c4ba8bb in core::future::from_generator::{impl#1}::poll<mint_client::wallet::tests::create_output::{async_block_env#0}> (self=..., cx=*mut ...) at /rustc/f1f721e64014863f41c1a386b04af04c2de25321/library/core/src/future/mod.
rs:91

```
</details>

**Comment by jkitman** [comment]
Looks like this issue was fixed by #137, lmk if you think otherwise.
